### ISSUE TITLE
Feature request: Can anyone implement an operator equivalent to tensorflow.multinomial?

### ISSUE THREAD

**Issue by WarBean** [apache/incubator-mxnet#3831]
Tensorflow use operator [multinomial](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/api_docs/python/functions_and_classes/shard3/tf.multinomial.md) for sampling inside computational graph. It is useful for schedule sampling training of RNN. Recent work on generating sentences by GAN ([seqGAN](https://github.com/LantaoYu/SeqGAN)) also relies on `tf.multinomial` in their implementation.

I suppose that without such operator, it's troublesome to integrate sampling into computation flow during **training phase**. So I try to implement it myself. However, after some research I found no way to perform discrete distribution sampling in cuda API. Here's a unsuccessful trail: ([Generating sample from a non uniform discrete distribution.](https://devtalk.nvidia.com/default/topic/838082/generating-sample-from-a-non-uniform-discrete-distribution-/?offset=2)) 

Off course I can use `mx.operator.CustomOp` to write a Numpy version, but it's the last choice for me.

**Comment by sxjscience** [comment]
@WarBean You can first sample values from Uniform(0, 1) and then find which region in the CDF contains the value. For example, assume your probability distribution is `prob`, you can do the sampling via `numpy.searchsorted(numpy.cumsum(prob[ind]), rng.rand())`

**Comment by WarBean** [comment]
@sxjscience After reading `sample_op.h` I have same idea on how to continue for a CPU version (something logic like `numpy.searchsorted(numpy.cumsum(prob), rng.rand())` as you mentioned) . But I'm not sure how to continue on for GPU version. More precisely, I don't know how to integrate cuda kernel api to **sample in parallel within a mini-batch**. Could you offer some direction on how to directly use cuda kernel (calling a function with `__device__` qualifier) in `Forward`, `Backward` function? Or there is some other approaches?

**Comment by sxjscience** [comment]
@WarBean You can directly call cuda kernels in the forward & backward functions. https://github.com/dmlc/mxnet/blob/master/src/operator/roi_pooling.cu is one example.

**Comment by WarBean** [comment]
@sxjscience Exactly what I need. I will try it.

**Comment by asmushetzel** [comment]
mxnet now (since a week ago)  has a lot of additional sampling operators, see sample_op.h/cc and multisample-op.h/cc
The latter version is one where parameters for the distributions are input tensors.
So far we implemented exponential/gamma/poisson/negative binomial. And just for CPUs.
Will look whether we can add CPU-support for multinomial easily. Getting them all on GPUs as well is also something we will look at.

**Comment by sxjscience** [comment]
@asmushetzel What's the progress now for the GPU support?

**Comment by asmushetzel** [comment]
Almost done. I hope we can get the pull request out by next week. This will also change the CPU implementations as we are using generic code that is equivalent on GPU & CPU. 
One thing: This is a thread specifically about multinomial distribution. This distribution was brought in independently by Eric some time ago and supports CPU/GPU already. So not sure why this issue here is still open. Shouldn't we close it?

**Comment by sxjscience** [comment]
OK, do we have a separate issue for the general sampling OPs? I'm going to close it as it's only about multinomial, which has been supported. https://mxnet.incubator.apache.org/api/python/symbol.html#mxnet.symbol.sample_multinomial

**Comment by asmushetzel** [comment]
We don't have a separate issue for general sampling. Feel free to open one, otherwise I will put you on cc on the upcoming pull request

**Comment by piiswrong** [comment]
We already have multinomial and it works on both cpu and gpu
